### PR TITLE
net: tcp: Process zero window probes when our recv_wnd == 0

### PR DESCRIPTION
**Pull request by pfalcon**
The IP stack drops any TCP segment which doesn't fit into our
receive window. However, we still must accept Zero Window Probe
segments, which are segments, usually with data length of 1, which
a peer sends to us after we stayed with zero window for some time.
In this case, we need to repeat an ACK with the old ack number.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

### CODE REVIEW

**Comment by pfalcon** [comment]
This addresses an issue with handling Zero Window Probes, identified in https://github.com/zephyrproject-rtos/zephyr/issues/7876. (Note: there may be further issues with ZWP handling, so it's not correct to say that this patch fixes that ticket.)

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -2044,6 +2045,14 @@ NET_CONN_CB(tcp_established)
 
 	data_len = net_pkt_appdatalen(pkt);
 	if (data_len > net_tcp_get_recv_wnd(context->tcp)) {
+		/* In case we have zero window, we should still accept
+		 * Zero Window Probes from peer, which per convention
+		 * come with len=1.
+		 */
+		if (data_len == 1 && net_tcp_get_recv_wnd(context->tcp) == 0) {
```
Comment (rlubos) [inline subsys/net/ip/tcp.c]: `net_tcp_get_recv_wnd(context->tcp) == 0` is redundant here. If  `data_len == 1` and we'e passed `if (data_len > net_tcp_get_recv_wnd(context->tcp))` current receive window size have to be 0.

Comment (pfalcon) [reply subsys/net/ip/tcp.c]: > net_tcp_get_recv_wnd(context->tcp) == 0 is redundant here. If data_len == 1 and we'e passed if (data_len > net_tcp_get_recv_wnd(context->tcp)) current receive window size have to be 0.

That's a little bit too-smartish, I'd say. Will remove the code, but will need to explain it in the comment.

**Comment by pfalcon** [comment]
Made the optimization @rlubos suggested, @jukkar, @rlubos, please review this small patch.

**Review by rlubos — APPROVED** [review]

**Review by jukkar — APPROVED** [review]
LGTM

**Pull request merged**